rclpy/rclpy/executors.py

```diff
@@ -347,56 +347,95 @@ def spin_once_until_future_complete(
         raise NotImplementedError()
 
     def _take_timer(self, tmr):
-        with tmr.handle:
-            tmr.handle.call_timer()
-        return ()
+        try:
+            with tmr.handle:
+                tmr.handle.call_timer()
+                return (True,)
+        except InvalidHandle:
+            # Timer is a Destroyable, which means that on __enter__ it can throw an
+            # InvalidHandle exception if the entity has already been destroyed.  Handle that here
+            # by just returning an empty argument, which means we will skip doing any real work
+            # in _execute_timer below
+            pass
+
+        return (False,)
```

Comment:
Feel free to ignore this. It seems to me like it might be simpler if `_take_*()` methods returned the coroutine to execute, otherwise they returned `None` if there's nothing to execute.

That is in `_make_handler` instead of this:

https://github.com/ros2/rclpy/blob/540b809b1b4d6fc064cfd392834123f713593be4/rclpy/rclpy/executors.py#L438-L446

It would be

```python3
call_coroutine = take_from_waitlist(entity)

entity._executor_event = False
gc.trigger() 

try:
    if call_coroutine:
        await call_coroutine()
finally:
```

Where the `_take_*` methods return the coroutine to execute when there is something to do, and `None` otherwhise.

```python3
# Return from `_take_timer` is execute method with no data
with tmr.handle:
    async def _execute():
        await await_or_execute(tmr.callback)
    return _execute
return None  # Or nothing since this is the default return

# ...

# Return from `_take_subscription` is execute method with message
with sub.handle:
    # ...
    async def _execute():
        await await_or_execute(sub.callback, msg_info)
    return _execute
```

Reply:
Yep, I really like this idea.  It makes the argument list to `make_handler` shorter, and concentrates all of the knowledge on how to take and execute each entity in one place.  I've now done this in 55d1fe6.

Since I think this is the right way to go, I'm going to take this out of draft now.  Another review is appreciated.